contracts/examples/ping-pong-egld/src/lib.rs

```diff
@@ -0,0 +1,168 @@
+#![no_std]
+#![allow(unused_attributes)]
+
+imports!();
+
+mod user_status;
+
+use user_status::UserStatus;
+
+#[elrond_wasm_derive::contract(PingPongImpl)]
+pub trait PingPong {
+	#[init]
+	fn init(
+		&self,
+		fixed_sum: &BigUint,
+		duration: u64,
+		beginning: Option<u64>,
+		max_funds: Option<BigUint>,
+	) {
+		self.set_fixed_sum(fixed_sum);
+		let computed_beginning = beginning.unwrap_or(self.get_block_nonce());
+		let deadline = computed_beginning + duration;
+		self.set_deadline(deadline);
+		self.set_beginning(computed_beginning);
+		self.set_max_funds(max_funds);
+	}
+
+	#[payable]
+	#[endpoint]
+	fn ping(&self, #[payment] payment: &BigUint) -> SCResult<()> {
+		require!(
+			payment == &self.get_fixed_sum(),
+			"the payment must match the fixed sum"
+		);
+
+		require!(
+			self.get_beginning() <= self.get_block_nonce(),
+			"smart contract not active yet"
+		);
+
+		require!(
+			self.get_max_funds()
+				.map_or(true, |max_funds| &self.get_sc_balance() + payment
+					> max_funds),
+			"smart contract full"
+		);
+
+		let caller = self.get_caller();
+		let user_status = self.get_user_status(&caller);
+		match user_status {
+			UserStatus::New => {
+				self.set_user_status(&caller, &UserStatus::Registered);
+				let new_index = self.get_size() + 1;
+				self.set_index(&caller, new_index);
+				self.set_address(new_index, &caller);
+				self.set_size(new_index);
+				Ok(())
+			},
+			UserStatus::Registered => {
+				sc_error!("can only ping once")
+			},
+			UserStatus::Withdrawn => {
+				sc_error!("already withdrawn")
+			},
+		}
+	}
+
+	fn pong_address(&self, caller: &Address) -> SCResult<()> {
+		require!(
+			self.get_block_nonce() >= self.get_deadline(),
+			"can't withdraw before deadline"
+		);
+
+		let user_status = self.get_user_status(&caller);
+		match user_status {
+			UserStatus::New => {
+				sc_error!("can't pong, never pinged")
+			},
+			UserStatus::Registered => {
+				self.set_user_status(&caller, &UserStatus::Withdrawn);
+				self.send_tx(&caller, &self.get_fixed_sum(), "pong");
+				Ok(())
+			},
+			UserStatus::Withdrawn => {
+				sc_error!("already withdrawn")
+			},
+		}
+	}
+
+	#[endpoint]
+	fn pong(&self) -> SCResult<()> {
+		let caller = self.get_caller();
+		self.pong_address(&caller)
+	}
+
+	#[endpoint]
+	fn pong_all(&self) -> SCResult<()> {
+		require!(
+			self.get_block_nonce() >= self.get_deadline(),
+			"can't withdraw before deadline"
+		);
+
+		let size = self.get_size();
+		for i in 1..=size {
+			let address = self.get_address(i);
+			self.pong_address(&address).ok();
```

Comment:
Regarding pong_all - it doesn't crash since when any user encounters an error since the SCResult returned by pong (pong_address / pong_by_user_id) in the loop is ignored (note that .ok() is called on the returned value so that the compiler doesn't complain the result is ignored).
I agree with the rest. I changed the mapping to use the user_id instead of the user address (using the naming as well).
The get_or_create_user function proved useful since it allowed me to simplify the code inside the ping method.